src/common/src/hash/key.rs

```diff
@@ -114,9 +114,11 @@ pub trait HashKey:
         data_chunk: &DataChunk,
         hash_codes: Vec<HashCode>,
     ) -> Vec<Self> {
+        let estimated_key_size = data_chunk.estimate_value_encoding_size(column_idxes);
+        // Construct serializers for each row.
         let mut serializers: Vec<Self::S> = hash_codes
             .into_iter()
-            .map(Self::S::from_hash_code)
+            .map(|hashcode| Self::S::from_hash_code(hashcode, estimated_key_size))
```

Comment:
I think it is not good enough, we can get each row's value encoding size like https://github.com/risingwavelabs/risingwave/pull/8924. And after that, we can avoid any grow of the vec